[SPARK-14014] [SQL] Replace existing catalog with SessionCatalog

andrewor14 opened the pull request:
## What changes were proposed in this pull request?

`SessionCatalog`, introduced in #11750, is a catalog that keeps track of temporary functions and tables, and delegates metastore operations to `ExternalCatalog`. This functionality overlaps a lot with the existing `analysis.Catalog`.

As of this commit, `SessionCatalog` and `ExternalCatalog` will no longer be dead code. There are still things that need to be done after this patch, namely:
- SPARK-14013: Properly implement temporary functions in `SessionCatalog`
- SPARK-13879: Decide which DDL/DML commands to support natively in Spark
- SPARK-?????: Implement the ones we do want to support through `SessionCatalog`.
- SPARK-?????: Merge SQL/HiveContext
## How was this patch tested?

This is largely a refactoring task so there are no new tests introduced. The particularly relevant tests are `SessionCatalogSuite` and `ExternalCatalogSuite`.

Commit:
Squashed commit of the following:

commit ad43a5ffdeeb881aaed8944971b63a27d1f4257f
Author: Andrew Or <andrew@databricks.com>
Date:   Wed Mar 16 14:35:02 2016 -0700

    Expand test scope + clean up test code

commit 08969cdcaf8196a30a3c879f956a8386fe400695
Author: Andrew Or <andrew@databricks.com>
Date:   Wed Mar 16 13:21:50 2016 -0700

    Fix tests

commit 6d9fa2f946ac93ebc95a9f25cf515fb0ea54b17c
Author: Andrew Or <andrew@databricks.com>
Date:   Wed Mar 16 12:31:52 2016 -0700

    Keep track of current database in SessionCatalog

    This allows us to not pass it into every single method like
    we used to before this commit.

commit ff1c2c4661986622e8071a39922e25033b3e62ab
Author: Andrew Or <andrew@databricks.com>
Date:   Tue Mar 15 19:42:22 2016 -0700

    Add TODO

commit 8c84dd803829ffcb8c82ee2f593ef58c3c5c94c9
Author: Andrew Or <andrew@databricks.com>
Date:   Tue Mar 15 19:41:30 2016 -0700

    Implement tests for functions

commit 3da16fb3473b750f13ffcbbb8aaf9a7de7292897
Author: Andrew Or <andrew@databricks.com>
Date:   Tue Mar 15 19:04:03 2016 -0700

    Implement tests for table partitions

commit 794744565269bb9ffb00f8d7a81d7b703251f956
Author: Andrew Or <andrew@databricks.com>
Date:   Tue Mar 15 18:52:30 2016 -0700

    Implement tests for databases and tables

commit 2f5121b43c938b2b585de0c3d80680c0ad5a8a7d
Author: Andrew Or <andrew@databricks.com>
Date:   Tue Mar 15 16:59:38 2016 -0700

    Fix infinite loop (woops)

commit d3f252d4d21b91a22dd7277f983f84daa56d65b5
Author: Andrew Or <andrew@databricks.com>
Date:   Tue Mar 15 16:12:55 2016 -0700

    Refactor CatalogTestCases to make methods accessible

commit caa4013e457a46ef0b8c3a2291cb375eb9064972
Author: Andrew Or <andrew@databricks.com>
Date:   Tue Mar 15 15:44:23 2016 -0700

    Clean up duplicate code in Table/FunctionIdentifier

commit 90ccdbb22bd8baf8caf839047148ebfd326b3593
Author: Andrew Or <andrew@databricks.com>
Date:   Tue Mar 15 15:33:30 2016 -0700

    Fix style

commit 5587a4995634af44ceecc9755165eb9a02bc0e5b
Author: Andrew Or <andrew@databricks.com>
Date:   Tue Mar 15 15:32:38 2016 -0700

    Implement SessionCatalog using ExternalCatalog

commit 196f7ce1b9cfdcd607e363be10716c2dec409bd2
Author: Andrew Or <andrew@databricks.com>
Date:   Tue Mar 15 14:39:22 2016 -0700

    Document and clean up function methods

commit 6d530a919c2f61e69d970625f77b99df5c93b019
Author: Andrew Or <andrew@databricks.com>
Date:   Tue Mar 15 14:38:50 2016 -0700

    Fix tests

commit 2118212a6b5314838d322169c756714d9670d9ac
Author: Andrew Or <andrew@databricks.com>
Date:   Tue Mar 15 14:33:20 2016 -0700

    Refactor CatalogFunction to use FunctionIdentifier

commit dd1fbaef9f53cb61cf726b95fe2bd1a845afa2c3
Author: Andrew Or <andrew@databricks.com>
Date:   Tue Mar 15 14:22:37 2016 -0700

    Refactor CatalogTable to use TableIdentifier

    This is a standalone commit such that in the future we can split
    it out into a separate patch if preferrable.

commit 39a153c1b5ac495766eed13c5bb5e5f1135a4e4f
Author: Andrew Or <andrew@databricks.com>
Date:   Tue Mar 15 13:53:42 2016 -0700

    Take into account current database in table methods

commit 5bf695c686d84df500b36713b2ef86226615f3c6
Author: Andrew Or <andrew@databricks.com>
Date:   Mon Mar 14 17:14:59 2016 -0700

    Do the same for functions and partitions

commit 1d12578708da845fe309d3aae1dcdadfee1dee89
Author: Andrew Or <andrew@databricks.com>
Date:   Mon Mar 14 16:27:11 2016 -0700

    Clean up table method signatures + add comments

commit 98c8a3b922168b843fe648664fc0e8ac2f930472
Author: Andrew Or <andrew@databricks.com>
Date:   Thu Mar 10 16:35:35 2016 -0800

    Merge in @yhuai's changes

Commit:
Refactor SQLContext etc. to take in ExternalCatalog

We need to be able to pass in ExternalCatalog in the constructor
of SQLContext and subclasses because these should be persistent
across sessions. Unfortunately without significant refactoring
in the HiveContext and TestHive code we cannot make this simple
change happen.

Commit:
Attempt to remove old catalog from SessionState

This failed because SessionCatalog does not implement
refreshTable. This is a bigger problem because SessionCatalog
has no notion of caching tables in the first place and so it
doesn't really make sense to implement refreshTable. More
refactoring involving HiveMetastoreCatalog is required to
make this work.

Commit:
Merge branch 'master' of github.com:apache/spark into use-session-catalog

Commit:
Fix style

Commit:
Replace all usages of analysis.Catalog

This commit deletes the trait analysis.Catalog and all of its
subclasses, with one notable exception: HiveMetastoreCatalog
is kept because a lot of existing functionality (like caching
data source tables) are still needed. All other occurrences
are now replaced with SessionCatalog.

Unfortunately, because HiveMetastoreCatalog is a massive
sprawl of unmaintainable code, there is no clean way to
integrate it nicely with the new HiveCatalog. The path of
least resistance, then, route previous usages of
HiveMetastoreCatalog through HiveCatalog. This requires
some whacky initialization order hacks because HMC takes
in HiveContext but HiveContext takes in HiveCatalog.

Commit:
Fix tests

The biggest change here is moving HiveMetastoreCatalog from
HiveCatalog (the external one) to HiveSessionCatalog (the session
specific one). This is needed because HMC depends on a lot of
session specific things for, e.g. creating data source tables.
This was failing tests that do things with multiple sessions,
i.e. HiveQuerySuite.

Commit:
Fix tests round 2

There were some issues with case sensitivity analysis and error
messages not being exactly as expected. The latter is now relaxed
where possible.

andrewor14 commented:
@yhuai @rxin

SparkQA commented:
**[Test build #53579 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53579/consoleFull)** for PR 11836 at commit [`5e16480`](https://github.com/apache/spark/commit/5e1648074ffb96f1b2104dc5ea3d78d25e505181).

SparkQA commented:
**[Test build #53579 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53579/consoleFull)** for PR 11836 at commit [`5e16480`](https://github.com/apache/spark/commit/5e1648074ffb96f1b2104dc5ea3d78d25e505181).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class SimpleAnalyzer(conf: CatalystConf)`
  * `class SessionCatalog(externalCatalog: ExternalCatalog, conf: CatalystConf) `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53579/
Test FAILed.

Commit:
Fix MiMa

Commit:
Merge branch 'master' of github.com:apache/spark into use-session-catalog

SparkQA commented:
**[Test build #53582 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53582/consoleFull)** for PR 11836 at commit [`c439280`](https://github.com/apache/spark/commit/c439280820a3478c45b64de8c605b0cc0f96e1a1).

Commit:
Minor fixes

Commit:
sessionState.sessionCatalog -> sessionState.catalog

Commit:
Fix tests round 3 (small round)

SparkQA commented:
**[Test build #53585 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53585/consoleFull)** for PR 11836 at commit [`34948a5`](https://github.com/apache/spark/commit/34948a5857453b9693113e57915a5d6d6e328fca).

SparkQA commented:
**[Test build #53587 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53587/consoleFull)** for PR 11836 at commit [`f089e2b`](https://github.com/apache/spark/commit/f089e2bebacc000ac65a0a14b1124c0c5a1e860c).

SparkQA commented:
**[Test build #53582 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53582/consoleFull)** for PR 11836 at commit [`c439280`](https://github.com/apache/spark/commit/c439280820a3478c45b64de8c605b0cc0f96e1a1).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53582/
Test FAILed.

SparkQA commented:
**[Test build #53587 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53587/consoleFull)** for PR 11836 at commit [`f089e2b`](https://github.com/apache/spark/commit/f089e2bebacc000ac65a0a14b1124c0c5a1e860c).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53587/
Test FAILed.

SparkQA commented:
**[Test build #53585 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53585/consoleFull)** for PR 11836 at commit [`34948a5`](https://github.com/apache/spark/commit/34948a5857453b9693113e57915a5d6d6e328fca).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53585/
Test FAILed.

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/InMemoryCatalog.scala:
```diff
@@ -52,37 +52,34 @@ class InMemoryCatalog extends ExternalCatalog {
-  }
-
-  private def existsPartition(db: String, table: String, spec: TablePartitionSpec): Boolean = {
+  private def partitionExists(db: String, table: String, spec: TablePartitionSpec): Boolean = {
```
in the future let's separate renames likes this from large prs.

rxin commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveCatalog.scala:
```diff
@@ -85,6 +85,13 @@ private[spark] class HiveCatalog(client: HiveClient) extends ExternalCatalog wit
+  /**
+   * Switch our client to one that belongs to the new session.
+   */
+  def newSession(newClient: HiveClient): this.type = {
```
i'm not sure about this one, but should we just consider creating the hive client in the ctor itself?

andrewor14 replied:
just talked to @yhuai offline. I don't think we need this. The whole reason why I did something like this is because `HiveClient` also has session-specific stuff, but as long as we use the root `HiveClient` (the first one) then we're fine.

rxin replied:
yea - if it is going to take a while feel free to do as an add-on pr. it'd be fix tests and great to merge this one asap.

andrewor14 replied:
already did it; only ~5 lines change

rxin commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveCatalog.scala:
```diff
@@ -182,13 +189,15 @@ private[spark] class HiveCatalog(client: HiveClient) extends ExternalCatalog wit
+  }
+
   override def listTables(db: String): Seq[String] = withClient {
-    requireDbExists(db)
```
why removing this?

andrewor14 replied:
it fails existing tests

yhuai replied:
Which test?

andrewor14 replied:
ListTablesSuite

yhuai replied:
What was the error?

andrewor14 replied:
It just assumes that listing tables of a non-existent database would work.

yhuai replied:
Ah I see. Our current master can allow you call `SHOW TABLES` with a db that does not actually exist and then we displays all temp tables. This behavior can cause confusion to users because they can think that the database exists. But, later when they call `use` to switch to that non-existent db, they will see an exception.

rxin replied:
seems like we should throw exception when db doesn't exist.

we could do that as a separate pr though

andrewor14 replied:
this was done in this patch itself.

rxin commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveContext.scala:
```diff
@@ -81,15 +83,31 @@ class HiveContext private[hive](
-    @transient private val metaHive: HiveClient,
-    isRootContext: Boolean)
-  extends SQLContext(sc, cacheManager, listener, isRootContext) with Logging {
+    @transient protected[hive] val executionHive: HiveClientImpl,
```
Why is there an executionHive?

andrewor14 replied:
what do you mean? The old code also had it.

rxin replied:
iirc it's used to find temp functions and the current database. we are doing those ourselves now, no?

andrewor14 replied:
IDK what we use it for but it's already there before so I just kept it. @yhuai

yhuai replied:
We have not changed temp function part. executionHive is still used when we look up Hive UDFs. The next step is to remove the usage of executionHive when we handle Hive UDFs. Then, we should be able to remove executionHive (So, HiveClient is only used to talk to real metastores.).

rxin commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveContext.scala:
```diff
@@ -666,6 +470,241 @@ private[hive] object HiveContext {
+  /**
+   * Configurations needed to create a [[HiveClient]].
+   */
+  private[hive] def hiveClientConfigurations(hiveconf: HiveConf): Map[String, String] = {
```
is the next step to move this into the hive catalog itself?

andrewor14 replied:
I don't understand. This is used to create a `HiveClient`, which happens before we create `HiveCatalog`.

rxin commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala:
```diff
@@ -0,0 +1,104 @@
+    client.setCurrentDatabase(db)
+  }
+
+  override def lookupRelation(name: TableIdentifier, alias: Option[String]): LogicalPlan = {
```
this should go away in the future right?

andrewor14 replied:
well we still need the logic in `HiveMetastoreCatalog`. It might be refactored and moved to the external catalog.

yhuai replied:
Eventually, we should not need to have a `HiveSessionCatalog`. For now, the main reason that we have it is because  `HiveMetastoreCatalog` is tightly coupled with both Hive and Spark SQL (e.g. the metadata cache's load method reads metadata from Hive metastore and then creates LogicalPlans directly).

(@andrewor14 correct me if I am missing anything)

rxin commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveQuerySuite.scala:
```diff
@@ -57,11 +58,14 @@ class HiveQuerySuite extends HiveComparisonTest with BeforeAndAfter {
-    Locale.setDefault(originalLocale)
-    sql("DROP TEMPORARY FUNCTION udtf_count2")
-    super.afterAll()
+    try {
```
good change!

rxin commented:
The changes look pretty good. We'd need to make Jenkins pass though.

Commit:
Merge branch 'master' of github.com:apache/spark into use-session-catalog

Commit:
Don't bother sessionizing HiveCatalog

Commit:
Fix tests (round 4) - ignored test in CliSuite

Note: This commit ignores a test in CliSuite. There a future
timed out and I investigated for like half an hour and could
not figure out why. It has something to do with the way we set
the current database and executing commands with "-e". This
will take a little longer to debug so I prefer to do that in
a separate patch.

andrewor14 commented:
The latest commit should pass tests. Note that I've ignored a test in `CliSuite` because I have no idea why the future timed out even after investigating for 30 minutes. It's an end-to-end test with almost no intermediate output so it's would much longer to debug the root cause. If that's the only failing test then I'd prefer to fix it after this patch gets merged so we unblock progress on the DDL stuff.

SparkQA commented:
**[Test build #53621 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53621/consoleFull)** for PR 11836 at commit [`4b37d7a`](https://github.com/apache/spark/commit/4b37d7aae3bdaaf61dba18d23dae2c7da9938a5f).

yhuai commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala:
```diff
@@ -31,17 +32,34 @@ import org.apache.spark.sql.catalyst.plans.logical.{LogicalPlan, SubqueryAlias}
+   * Format table name, taking into account case sensitivity.
+   */
+  protected[this] def formatTableName(name: String): String = {
+    if (conf.caseSensitiveAnalysis) name else name.toLowerCase
```
Just a note at here. Hive metastore is always case insensitive. So, the case sensitivity setting is mainly for temp tables and temp functions.

yhuai replied:
Later, it will be good to use this to handle other db name as well for the consistency reason (it will not actually have any effect right now though).

yhuai commented:
Changes look good to me.

SparkQA commented:
**[Test build #53621 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53621/consoleFull)** for PR 11836 at commit [`4b37d7a`](https://github.com/apache/spark/commit/4b37d7aae3bdaaf61dba18d23dae2c7da9938a5f).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53621/
Test FAILed.

andrewor14 commented:
ahhh that's a lot more failed tests. I'll investigate tonight.

Commit:
Merge branch 'master' of github.com:apache/spark into use-session-catalog

Commit:
Clear temp tables after each suite

SparkQA commented:
**[Test build #53702 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53702/consoleFull)** for PR 11836 at commit [`52e0273`](https://github.com/apache/spark/commit/52e027367dc03fcdec1aab7792f6e332e16f14a7).

Commit:
Require DB exists before showing tables on them

SparkQA commented:
**[Test build #53703 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53703/consoleFull)** for PR 11836 at commit [`19750d7`](https://github.com/apache/spark/commit/19750d74230e1839c0b678be946b79e5afe43261).

SparkQA commented:
**[Test build #53703 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53703/consoleFull)** for PR 11836 at commit [`19750d7`](https://github.com/apache/spark/commit/19750d74230e1839c0b678be946b79e5afe43261).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53703/
Test FAILed.

Commit:
Fix tests

SparkQA commented:
**[Test build #53707 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53707/consoleFull)** for PR 11836 at commit [`561ca3c`](https://github.com/apache/spark/commit/561ca3ce16d4e4fbd1bc77c4484cefeed45f9f7d).

Commit:
Fix MultiDatabaseSuite

SparkQA commented:
**[Test build #53713 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53713/consoleFull)** for PR 11836 at commit [`b9de78c`](https://github.com/apache/spark/commit/b9de78c980bca3738cb493056326cab1c81ed343).

SparkQA commented:
**[Test build #53702 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53702/consoleFull)** for PR 11836 at commit [`52e0273`](https://github.com/apache/spark/commit/52e027367dc03fcdec1aab7792f6e332e16f14a7).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53702/
Test FAILed.

SparkQA commented:
**[Test build #53707 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53707/consoleFull)** for PR 11836 at commit [`561ca3c`](https://github.com/apache/spark/commit/561ca3ce16d4e4fbd1bc77c4484cefeed45f9f7d).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53707/
Test FAILed.

andrewor14 commented:
retest this please

SparkQA commented:
**[Test build #53713 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53713/consoleFull)** for PR 11836 at commit [`b9de78c`](https://github.com/apache/spark/commit/b9de78c980bca3738cb493056326cab1c81ed343).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53713/
Test FAILed.

yhuai commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveContext.scala:
```diff
@@ -666,6 +468,241 @@ private[hive] object HiveContext {
+    val hiveMetastoreVersion = HiveContext.hiveMetastoreVersion(sqlConf)
+    val hiveMetastoreJars = HiveContext.hiveMetastoreJars(sqlConf)
+    val hiveMetastoreSharedPrefixes = HiveContext.hiveMetastoreSharedPrefixes(sqlConf)
+    val hiveMetastoreBarrierPrefixes = HiveContext.hiveMetastoreBarrierPrefixes(sqlConf)
```
When we create the sqlConf in this method, that conf object only hold default values and do not have settings of these four keys. We need to populate the sqlConf by copy sql related keys from the spark conf to it like what we do at https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala#L206-L228. Then, we will get the correct values of these four keys. 

This will help us fix `HiveUDFSuite`.

Commit:
Merge branch 'master' of github.com:apache/spark into use-session-catalog

Commit:
Fix HiveUDFSuite + add tests

The problem was that the metadataHive didn't get any of the
spark.sql.* confs, so the barrier prefixes weren't actually set.
Thanks to @yhuai for uncovering this.

SparkQA commented:
**[Test build #53792 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53792/consoleFull)** for PR 11836 at commit [`4133d3f`](https://github.com/apache/spark/commit/4133d3f64747987728a0db227d32d3001e846996).

SparkQA commented:
**[Test build #53792 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53792/consoleFull)** for PR 11836 at commit [`4133d3f`](https://github.com/apache/spark/commit/4133d3f64747987728a0db227d32d3001e846996).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53792/
Test FAILed.

Commit:
Fix HiveCompatibilitySuite?

The issue is that after each test we only set the current
database in Hive but not the one in SessionCatalog. This means
the next test will create a table in the default database (since
we just pass CREATE TABLE commands to hive currently) but try
to resolve it in a database left over from a previous test.

Commit:
Fix CliSuite

We were expecting an "OK" that never came. This test is way
to specific anyway and is super brittle. It's also better to
alawys set the current database through the catalog so we don't
end up with mismatched current databases between Spark and Hive.

Commit:
Merge branch 'master' of github.com:apache/spark into use-session-catalog

SparkQA commented:
**[Test build #53820 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53820/consoleFull)** for PR 11836 at commit [`98751cc`](https://github.com/apache/spark/commit/98751ccf97345883310819655139ead59e877c07).

Commit:
Fix HiveQuerySuite?

Every time we called TestHive.reset() we created a new temp
directory for derby, and then we would go ahead and override
the old one in the same TestHiveContext. This fails tests that
use multiple sessions for some reason. Setting the same confs
in metadataHive whenever we call reset() seems unnecessary,
so I removed it.

SparkQA commented:
**[Test build #53823 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53823/consoleFull)** for PR 11836 at commit [`16a54ba`](https://github.com/apache/spark/commit/16a54bad76a8297f15417ba931d20c4d86092c84).

SparkQA commented:
**[Test build #53820 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53820/consoleFull)** for PR 11836 at commit [`98751cc`](https://github.com/apache/spark/commit/98751ccf97345883310819655139ead59e877c07).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class LogisticRegressionModel(JavaModel, JavaMLWritable, JavaMLReadable):`
  * `class NaiveBayesModel(JavaModel, JavaMLWritable, JavaMLReadable):`
  * `class KMeansModel(JavaModel, JavaMLWritable, JavaMLReadable):`
  * `class Binarizer(JavaTransformer, HasInputCol, HasOutputCol, JavaMLReadable, JavaMLWritable):`
  * `class Bucketizer(JavaTransformer, HasInputCol, HasOutputCol, JavaMLReadable, JavaMLWritable):`
  * `class CountVectorizer(JavaEstimator, HasInputCol, HasOutputCol, JavaMLReadable, JavaMLWritable):`
  * `class CountVectorizerModel(JavaModel, JavaMLReadable, JavaMLWritable):`
  * `class DCT(JavaTransformer, HasInputCol, HasOutputCol, JavaMLReadable, JavaMLWritable):`
  * `class ElementwiseProduct(JavaTransformer, HasInputCol, HasOutputCol, JavaMLReadable,`
  * `class HashingTF(JavaTransformer, HasInputCol, HasOutputCol, HasNumFeatures, JavaMLReadable,`
  * `class IDF(JavaEstimator, HasInputCol, HasOutputCol, JavaMLReadable, JavaMLWritable):`
  * `class IDFModel(JavaModel, JavaMLReadable, JavaMLWritable):`
  * `class MaxAbsScaler(JavaEstimator, HasInputCol, HasOutputCol, JavaMLReadable, JavaMLWritable):`
  * `class MaxAbsScalerModel(JavaModel, JavaMLReadable, JavaMLWritable):`
  * `class MinMaxScaler(JavaEstimator, HasInputCol, HasOutputCol, JavaMLReadable, JavaMLWritable):`
  * `class MinMaxScalerModel(JavaModel, JavaMLReadable, JavaMLWritable):`
  * `class NGram(JavaTransformer, HasInputCol, HasOutputCol, JavaMLReadable, JavaMLWritable):`
  * `class Normalizer(JavaTransformer, HasInputCol, HasOutputCol, JavaMLReadable, JavaMLWritable):`
  * `class OneHotEncoder(JavaTransformer, HasInputCol, HasOutputCol, JavaMLReadable, JavaMLWritable):`
  * `class PolynomialExpansion(JavaTransformer, HasInputCol, HasOutputCol, JavaMLReadable,`
  * `class QuantileDiscretizer(JavaEstimator, HasInputCol, HasOutputCol, HasSeed, JavaMLReadable,`
  * `class RegexTokenizer(JavaTransformer, HasInputCol, HasOutputCol, JavaMLReadable, JavaMLWritable):`
  * `class SQLTransformer(JavaTransformer, JavaMLReadable, JavaMLWritable):`
  * `class StandardScaler(JavaEstimator, HasInputCol, HasOutputCol, JavaMLReadable, JavaMLWritable):`
  * `class StandardScalerModel(JavaModel, JavaMLReadable, JavaMLWritable):`
  * `class StringIndexer(JavaEstimator, HasInputCol, HasOutputCol, HasHandleInvalid, JavaMLReadable,`
  * `class StringIndexerModel(JavaModel, JavaMLReadable, JavaMLWritable):`
  * `class IndexToString(JavaTransformer, HasInputCol, HasOutputCol, JavaMLReadable, JavaMLWritable):`
  * `class StopWordsRemover(JavaTransformer, HasInputCol, HasOutputCol, JavaMLReadable, JavaMLWritable):`
  * `class Tokenizer(JavaTransformer, HasInputCol, HasOutputCol, JavaMLReadable, JavaMLWritable):`
  * `class VectorAssembler(JavaTransformer, HasInputCols, HasOutputCol, JavaMLReadable, JavaMLWritable):`
  * `class VectorIndexer(JavaEstimator, HasInputCol, HasOutputCol, JavaMLReadable, JavaMLWritable):`
  * `class VectorIndexerModel(JavaModel, JavaMLReadable, JavaMLWritable):`
  * `class VectorSlicer(JavaTransformer, HasInputCol, HasOutputCol, JavaMLReadable, JavaMLWritable):`
  * `class Word2VecModel(JavaModel, JavaMLReadable, JavaMLWritable):`
  * `class PCA(JavaEstimator, HasInputCol, HasOutputCol, JavaMLReadable, JavaMLWritable):`
  * `class PCAModel(JavaModel, JavaMLReadable, JavaMLWritable):`
  * `class RFormula(JavaEstimator, HasFeaturesCol, HasLabelCol, JavaMLReadable, JavaMLWritable):`
  * `class RFormulaModel(JavaModel, JavaMLReadable, JavaMLWritable):`
  * `class ChiSqSelector(JavaEstimator, HasFeaturesCol, HasOutputCol, HasLabelCol, JavaMLReadable,`
  * `class ChiSqSelectorModel(JavaModel, JavaMLReadable, JavaMLWritable):`
  * `class PipelineMLWriter(JavaMLWriter):`
  * `class Pipeline(Estimator, MLReadable, MLWritable):`
  * `class PipelineModelMLWriter(JavaMLWriter):`
  * `class PipelineModel(Model, MLReadable, MLWritable):`
  * `class ALSModel(JavaModel, JavaMLWritable, JavaMLReadable):`
  * `class LinearRegressionModel(JavaModel, JavaMLWritable, JavaMLReadable):`
  * `class IsotonicRegressionModel(JavaModel, JavaMLWritable, JavaMLReadable):`
  * `class AFTSurvivalRegressionModel(JavaModel, JavaMLWritable, JavaMLReadable):`
  * `class MLWriter(object):`
  * `class JavaMLWriter(MLWriter):`
  * `class JavaMLWritable(MLWritable):`
  * `class MLReader(object):`
  * `class JavaMLReader(MLReader):`
  * `class JavaMLReadable(MLReadable):`
  * `  implicit class StringToColumn(val sc: StringContext) `
  * `class HDFSMetadataLog[T: ClassTag](sqlContext: SQLContext, path: String)`
  * `class StreamProgress(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53820/
Test FAILed.

SparkQA commented:
**[Test build #53823 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53823/consoleFull)** for PR 11836 at commit [`16a54ba`](https://github.com/apache/spark/commit/16a54bad76a8297f15417ba931d20c4d86092c84).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53823/
Test FAILed.

Commit:
Ignore new test for now...

SparkQA commented:
**[Test build #53848 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53848/consoleFull)** for PR 11836 at commit [`3439dc2`](https://github.com/apache/spark/commit/3439dc216dbaf6b7ab23246d36d9ba4bf52847ed).

andrewor14 commented:
retest this please

SparkQA commented:
**[Test build #53849 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53849/consoleFull)** for PR 11836 at commit [`3439dc2`](https://github.com/apache/spark/commit/3439dc216dbaf6b7ab23246d36d9ba4bf52847ed).

SparkQA commented:
**[Test build #53849 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53849/consoleFull)** for PR 11836 at commit [`3439dc2`](https://github.com/apache/spark/commit/3439dc216dbaf6b7ab23246d36d9ba4bf52847ed).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53849/
Test FAILed.

andrewor14 commented:
Hm, something weird is going on with the tests. The latest this please try again.

Commit:
Fix HiveContextSuite?

SparkQA commented:
**[Test build #53856 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53856/consoleFull)** for PR 11836 at commit [`e552558`](https://github.com/apache/spark/commit/e5525581d6b92b4306076fae75a7321fe346e650).

SparkQA commented:
**[Test build #53848 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53848/consoleFull)** for PR 11836 at commit [`3439dc2`](https://github.com/apache/spark/commit/3439dc216dbaf6b7ab23246d36d9ba4bf52847ed).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53848/
Test FAILed.

rxin commented:
ah only one test failure now 

```
[info] - HiveContext can access `spark.sql.*` configs *** FAILED *** (5 seconds, 654 milliseconds)
[info]   java.lang.reflect.InvocationTargetException:
[info]   at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
[info]   at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:57)
[info]   at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
[info]   at java.lang.reflect.Constructor.newInstance(Constructor.java:526)
[info]   at org.apache.spark.sql.hive.client.IsolatedClientLoader.createClient(IsolatedClientLoader.scala:257)
[info]   at org.apache.spark.sql.hive.HiveContext$.newClientForMetadata(HiveContext.scala:704)
[info]   at org.apache.spark.sql.hive.HiveContext$.org$apache$spark$sql$hive$HiveContext$$newClientForMetadata(HiveContext.scala:601)
[info]   at org.apache.spark.sql.hive.HiveContext.<init>(HiveContext.scala:108)
[info]   at org.apache.spark.sql.hive.HiveContextSuite$$anonfun$1.apply$mcV$sp(HiveContextSuite.scala:29)
```

yhuai commented:
test this please

SparkQA commented:
**[Test build #53890 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53890/consoleFull)** for PR 11836 at commit [`e552558`](https://github.com/apache/spark/commit/e5525581d6b92b4306076fae75a7321fe346e650).

yhuai commented:
I tried `hive/test` locally and I got the following error at `HiveContextSuite`.

```
21:43:03.057 ERROR DataNucleus.Datastore.Schema: Failed initialising database.
Unable to open a test connection to the given database. JDBC url = jdbc:derby:;databaseName=metastore_db;create=true, username = APP. Terminating connection pool (set lazyInit to true if you expect to start your database after your app). Original Exception: ------
java.sql.SQLException: Failed to start database 'metastore_db' with class loader org.apache.spark.sql.hive.client.IsolatedClientLoader$$anon$1@25b50060, see the next exception for details.
	at org.apache.derby.impl.jdbc.SQLExceptionFactory40.getSQLException(Unknown Source)
	at org.apache.derby.impl.jdbc.Util.newEmbedSQLException(Unknown Source)
	at org.apache.derby.impl.jdbc.Util.seeNextException(Unknown Source)
	at org.apache.derby.impl.jdbc.EmbedConnection.bootDatabase(Unknown Source)
	at org.apache.derby.impl.jdbc.EmbedConnection.<init>(Unknown Source)
	at org.apache.derby.impl.jdbc.EmbedConnection40.<init>(Unknown Source)
	at org.apache.derby.jdbc.Driver40.getNewEmbedConnection(Unknown Source)
	at org.apache.derby.jdbc.InternalDriver.connect(Unknown Source)
	at org.apache.derby.jdbc.Driver20.connect(Unknown Source)
	at org.apache.derby.jdbc.AutoloadedDriver.connect(Unknown Source)
	at java.sql.DriverManager.getConnection(DriverManager.java:571)
	at java.sql.DriverManager.getConnection(DriverManager.java:187)
	at com.jolbox.bonecp.BoneCP.obtainRawInternalConnection(BoneCP.java:361)
	at com.jolbox.bonecp.BoneCP.<init>(BoneCP.java:416)
	at com.jolbox.bonecp.BoneCPDataSource.getConnection(BoneCPDataSource.java:120)
	at org.datanucleus.store.rdbms.ConnectionFactoryImpl$ManagedConnectionImpl.getConnection(ConnectionFactoryImpl.java:501)
	at org.datanucleus.store.rdbms.RDBMSStoreManager.<init>(RDBMSStoreManager.java:298)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:57)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:526)
	at org.datanucleus.plugin.NonManagedPluginRegistry.createExecutableExtension(NonManagedPluginRegistry.java:631)
	at org.datanucleus.plugin.PluginManager.createExecutableExtension(PluginManager.java:301)
	at org.datanucleus.NucleusContext.createStoreManagerForProperties(NucleusContext.java:1187)
	at org.datanucleus.NucleusContext.initialise(NucleusContext.java:356)
	at org.datanucleus.api.jdo.JDOPersistenceManagerFactory.freezeConfiguration(JDOPersistenceManagerFactory.java:775)
	at org.datanucleus.api.jdo.JDOPersistenceManagerFactory.createPersistenceManagerFactory(JDOPersistenceManagerFactory.java:333)
	at org.datanucleus.api.jdo.JDOPersistenceManagerFactory.getPersistenceManagerFactory(JDOPersistenceManagerFactory.java:202)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at javax.jdo.JDOHelper$16.run(JDOHelper.java:1965)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.jdo.JDOHelper.invoke(JDOHelper.java:1960)
	at javax.jdo.JDOHelper.invokeGetPersistenceManagerFactoryOnImplementation(JDOHelper.java:1166)
	at javax.jdo.JDOHelper.getPersistenceManagerFactory(JDOHelper.java:808)
	at javax.jdo.JDOHelper.getPersistenceManagerFactory(JDOHelper.java:701)
	at org.apache.hadoop.hive.metastore.ObjectStore.getPMF(ObjectStore.java:365)
	at org.apache.hadoop.hive.metastore.ObjectStore.getPersistenceManager(ObjectStore.java:394)
	at org.apache.hadoop.hive.metastore.ObjectStore.initialize(ObjectStore.java:291)
	at org.apache.hadoop.hive.metastore.ObjectStore.setConf(ObjectStore.java:258)
	at org.apache.hadoop.util.ReflectionUtils.setConf(ReflectionUtils.java:73)
	at org.apache.hadoop.util.ReflectionUtils.newInstance(ReflectionUtils.java:133)
	at org.apache.hadoop.hive.metastore.RawStoreProxy.<init>(RawStoreProxy.java:57)
	at org.apache.hadoop.hive.metastore.RawStoreProxy.getProxy(RawStoreProxy.java:66)
	at org.apache.hadoop.hive.metastore.HiveMetaStore$HMSHandler.newRawStore(HiveMetaStore.java:593)
	at org.apache.hadoop.hive.metastore.HiveMetaStore$HMSHandler.getMS(HiveMetaStore.java:571)
	at org.apache.hadoop.hive.metastore.HiveMetaStore$HMSHandler.createDefaultDB(HiveMetaStore.java:620)
	at org.apache.hadoop.hive.metastore.HiveMetaStore$HMSHandler.init(HiveMetaStore.java:461)
	at org.apache.hadoop.hive.metastore.RetryingHMSHandler.<init>(RetryingHMSHandler.java:66)
	at org.apache.hadoop.hive.metastore.RetryingHMSHandler.getProxy(RetryingHMSHandler.java:72)
	at org.apache.hadoop.hive.metastore.HiveMetaStore.newRetryingHMSHandler(HiveMetaStore.java:5762)
	at org.apache.hadoop.hive.metastore.HiveMetaStoreClient.<init>(HiveMetaStoreClient.java:199)
	at org.apache.hadoop.hive.ql.metadata.SessionHiveMetaStoreClient.<init>(SessionHiveMetaStoreClient.java:74)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:57)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:526)
	at org.apache.hadoop.hive.metastore.MetaStoreUtils.newInstance(MetaStoreUtils.java:1521)
	at org.apache.hadoop.hive.metastore.RetryingMetaStoreClient.<init>(RetryingMetaStoreClient.java:86)
	at org.apache.hadoop.hive.metastore.RetryingMetaStoreClient.getProxy(RetryingMetaStoreClient.java:132)
	at org.apache.hadoop.hive.metastore.RetryingMetaStoreClient.getProxy(RetryingMetaStoreClient.java:104)
	at org.apache.hadoop.hive.ql.metadata.Hive.createMetaStoreClient(Hive.java:3005)
	at org.apache.hadoop.hive.ql.metadata.Hive.getMSC(Hive.java:3024)
	at org.apache.hadoop.hive.ql.metadata.Hive.getAllDatabases(Hive.java:1234)
	at org.apache.hadoop.hive.ql.metadata.Hive.reloadFunctions(Hive.java:174)
	at org.apache.hadoop.hive.ql.metadata.Hive.<clinit>(Hive.java:166)
	at org.apache.hadoop.hive.ql.session.SessionState.start(SessionState.java:503)
	at org.apache.spark.sql.hive.client.HiveClientImpl.<init>(HiveClientImpl.scala:140)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:57)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:526)
	at org.apache.spark.sql.hive.client.IsolatedClientLoader.createClient(IsolatedClientLoader.scala:257)
	at org.apache.spark.sql.hive.HiveContext$.newClientForMetadata(HiveContext.scala:704)
	at org.apache.spark.sql.hive.HiveContext$.org$apache$spark$sql$hive$HiveContext$$newClientForMetadata(HiveContext.scala:601)
	at org.apache.spark.sql.hive.HiveContext.<init>(HiveContext.scala:108)
	at org.apache.spark.sql.hive.HiveContextSuite$$anonfun$1.apply$mcV$sp(HiveContextSuite.scala:29)
	at org.apache.spark.sql.hive.HiveContextSuite$$anonfun$1.apply(HiveContextSuite.scala:26)
	at org.apache.spark.sql.hive.HiveContextSuite$$anonfun$1.apply(HiveContextSuite.scala:26)
	at org.scalatest.Transformer$$anonfun$apply$1.apply$mcV$sp(Transformer.scala:22)
	at org.scalatest.OutcomeOf$class.outcomeOf(OutcomeOf.scala:85)
	at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
	at org.scalatest.Transformer.apply(Transformer.scala:22)
	at org.scalatest.Transformer.apply(Transformer.scala:20)
	at org.scalatest.FunSuiteLike$$anon$1.apply(FunSuiteLike.scala:166)
	at org.apache.spark.SparkFunSuite.withFixture(SparkFunSuite.scala:56)
	at org.scalatest.FunSuiteLike$class.invokeWithFixture$1(FunSuiteLike.scala:163)
	at org.scalatest.FunSuiteLike$$anonfun$runTest$1.apply(FunSuiteLike.scala:175)
	at org.scalatest.FunSuiteLike$$anonfun$runTest$1.apply(FunSuiteLike.scala:175)
	at org.scalatest.SuperEngine.runTestImpl(Engine.scala:306)
	at org.scalatest.FunSuiteLike$class.runTest(FunSuiteLike.scala:175)
	at org.scalatest.FunSuite.runTest(FunSuite.scala:1555)
	at org.scalatest.FunSuiteLike$$anonfun$runTests$1.apply(FunSuiteLike.scala:208)
	at org.scalatest.FunSuiteLike$$anonfun$runTests$1.apply(FunSuiteLike.scala:208)
	at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:413)
	at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:401)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:401)
	at org.scalatest.SuperEngine.org$scalatest$SuperEngine$$runTestsInBranch(Engine.scala:396)
	at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:483)
	at org.scalatest.FunSuiteLike$class.runTests(FunSuiteLike.scala:208)
	at org.scalatest.FunSuite.runTests(FunSuite.scala:1555)
	at org.scalatest.Suite$class.run(Suite.scala:1424)
	at org.scalatest.FunSuite.org$scalatest$FunSuiteLike$$super$run(FunSuite.scala:1555)
	at org.scalatest.FunSuiteLike$$anonfun$run$1.apply(FunSuiteLike.scala:212)
	at org.scalatest.FunSuiteLike$$anonfun$run$1.apply(FunSuiteLike.scala:212)
	at org.scalatest.SuperEngine.runImpl(Engine.scala:545)
	at org.scalatest.FunSuiteLike$class.run(FunSuiteLike.scala:212)
	at org.apache.spark.SparkFunSuite.org$scalatest$BeforeAndAfterAll$$super$run(SparkFunSuite.scala:28)
	at org.scalatest.BeforeAndAfterAll$class.liftedTree1$1(BeforeAndAfterAll.scala:257)
	at org.scalatest.BeforeAndAfterAll$class.run(BeforeAndAfterAll.scala:256)
	at org.apache.spark.SparkFunSuite.run(SparkFunSuite.scala:28)
	at org.scalatest.tools.Framework.org$scalatest$tools$Framework$$runSuite(Framework.scala:357)
	at org.scalatest.tools.Framework$ScalaTestTask.execute(Framework.scala:502)
	at sbt.ForkMain$Run$2.call(ForkMain.java:296)
	at sbt.ForkMain$Run$2.call(ForkMain.java:286)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.sql.SQLException: Failed to start database 'metastore_db' with class loader org.apache.spark.sql.hive.client.IsolatedClientLoader$$anon$1@25b50060, see the next exception for details.
	at org.apache.derby.impl.jdbc.SQLExceptionFactory.getSQLException(Unknown Source)
	at org.apache.derby.impl.jdbc.SQLExceptionFactory40.wrapArgsForTransportAcrossDRDA(Unknown Source)
	... 122 more
Caused by: java.sql.SQLException: Another instance of Derby may have already booted the database /Users/yhuai/Projects/Spark/yin-spark-2/sql/hive/metastore_db.
	at org.apache.derby.impl.jdbc.SQLExceptionFactory.getSQLException(Unknown Source)
	at org.apache.derby.impl.jdbc.SQLExceptionFactory40.wrapArgsForTransportAcrossDRDA(Unknown Source)
	at org.apache.derby.impl.jdbc.SQLExceptionFactory40.getSQLException(Unknown Source)
	at org.apache.derby.impl.jdbc.Util.generateCsSQLException(Unknown Source)
	... 119 more
Caused by: ERROR XSDB6: Another instance of Derby may have already booted the database /Users/yhuai/Projects/Spark/yin-spark-2/sql/hive/metastore_db.
	at org.apache.derby.iapi.error.StandardException.newException(Unknown Source)
	at org.apache.derby.impl.store.raw.data.BaseDataFileFactory.privGetJBMSLockOnDB(Unknown Source)
	at org.apache.derby.impl.store.raw.data.BaseDataFileFactory.run(Unknown Source)
	at java.security.AccessController.doPrivileged(Native Method)
	at org.apache.derby.impl.store.raw.data.BaseDataFileFactory.getJBMSLockOnDB(Unknown Source)
	at org.apache.derby.impl.store.raw.data.BaseDataFileFactory.boot(Unknown Source)
	at org.apache.derby.impl.services.monitor.BaseMonitor.boot(Unknown Source)
	at org.apache.derby.impl.services.monitor.TopService.bootModule(Unknown Source)
	at org.apache.derby.impl.services.monitor.BaseMonitor.startModule(Unknown Source)
	at org.apache.derby.iapi.services.monitor.Monitor.bootServiceModule(Unknown Source)
	at org.apache.derby.impl.store.raw.RawStore.boot(Unknown Source)
	at org.apache.derby.impl.services.monitor.BaseMonitor.boot(Unknown Source)
	at org.apache.derby.impl.services.monitor.TopService.bootModule(Unknown Source)
	at org.apache.derby.impl.services.monitor.BaseMonitor.startModule(Unknown Source)
	at org.apache.derby.iapi.services.monitor.Monitor.bootServiceModule(Unknown Source)
	at org.apache.derby.impl.store.access.RAMAccessManager.boot(Unknown Source)
	at org.apache.derby.impl.services.monitor.BaseMonitor.boot(Unknown Source)
	at org.apache.derby.impl.services.monitor.TopService.bootModule(Unknown Source)
	at org.apache.derby.impl.services.monitor.BaseMonitor.startModule(Unknown Source)
	at org.apache.derby.iapi.services.monitor.Monitor.bootServiceModule(Unknown Source)
	at org.apache.derby.impl.db.BasicDatabase.bootStore(Unknown Source)
	at org.apache.derby.impl.db.BasicDatabase.boot(Unknown Source)
	at org.apache.derby.impl.services.monitor.BaseMonitor.boot(Unknown Source)
	at org.apache.derby.impl.services.monitor.TopService.bootModule(Unknown Source)
	at org.apache.derby.impl.services.monitor.BaseMonitor.bootService(Unknown Source)
	at org.apache.derby.impl.services.monitor.BaseMonitor.startProviderService(Unknown Source)
	at org.apache.derby.impl.services.monitor.BaseMonitor.findProviderAndStartService(Unknown Source)
	at org.apache.derby.impl.services.monitor.BaseMonitor.startPersistentService(Unknown Source)
	at org.apache.derby.iapi.services.monitor.Monitor.startPersistentService(Unknown Source)
	... 119 more
------
```

yhuai commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveContextSuite.scala:
```diff
@@ -0,0 +1,38 @@
+  test("HiveContext can access `spark.sql.*` configs") {
+    // Avoid creating another SparkContext in the same JVM
+    val sc = TestHive.sparkContext
+    val hiveContext = new HiveContext(sc)
```
Since there is already a global singleton, this probably will not work because of the derby connection limitation (inside the JVM, we can have a single connection to a given derby).

andrewor14 replied:
using `TestHive` doesn't work either. I'm just going to ignore the test again...

SparkQA commented:
**[Test build #53856 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53856/consoleFull)** for PR 11836 at commit [`e552558`](https://github.com/apache/spark/commit/e5525581d6b92b4306076fae75a7321fe346e650).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53856/
Test FAILed.

Commit:
Revert "Fix HiveContextSuite?"

This reverts commit e5525581d6b92b4306076fae75a7321fe346e650.

andrewor14 commented:
OK, now this should pass tests...

SparkQA commented:
**[Test build #2662 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2662/consoleFull)** for PR 11836 at commit [`5ea8469`](https://github.com/apache/spark/commit/5ea8469aafd347a7d1e69077de8d31a8f0167b25).

SparkQA commented:
**[Test build #2663 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2663/consoleFull)** for PR 11836 at commit [`5ea8469`](https://github.com/apache/spark/commit/5ea8469aafd347a7d1e69077de8d31a8f0167b25).

SparkQA commented:
**[Test build #2664 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2664/consoleFull)** for PR 11836 at commit [`5ea8469`](https://github.com/apache/spark/commit/5ea8469aafd347a7d1e69077de8d31a8f0167b25).

SparkQA commented:
**[Test build #2665 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2665/consoleFull)** for PR 11836 at commit [`5ea8469`](https://github.com/apache/spark/commit/5ea8469aafd347a7d1e69077de8d31a8f0167b25).

SparkQA commented:
**[Test build #53900 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53900/consoleFull)** for PR 11836 at commit [`5ea8469`](https://github.com/apache/spark/commit/5ea8469aafd347a7d1e69077de8d31a8f0167b25).

yhuai commented:
@andrewor14 Let's create a jira for that ignored test :)

SparkQA commented:
**[Test build #53890 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53890/consoleFull)** for PR 11836 at commit [`e552558`](https://github.com/apache/spark/commit/e5525581d6b92b4306076fae75a7321fe346e650).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53890/
Test FAILed.

rxin commented:
oh shit - conflict.

Commit:
Merge branch 'master' of github.com:apache/spark into use-session-catalog

SparkQA commented:
**[Test build #2664 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2664/consoleFull)** for PR 11836 at commit [`5ea8469`](https://github.com/apache/spark/commit/5ea8469aafd347a7d1e69077de8d31a8f0167b25).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

rxin commented:
looks really close ...

SparkQA commented:
**[Test build #2662 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2662/consoleFull)** for PR 11836 at commit [`5ea8469`](https://github.com/apache/spark/commit/5ea8469aafd347a7d1e69077de8d31a8f0167b25).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
**[Test build #2665 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2665/consoleFull)** for PR 11836 at commit [`5ea8469`](https://github.com/apache/spark/commit/5ea8469aafd347a7d1e69077de8d31a8f0167b25).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
**[Test build #53900 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53900/consoleFull)** for PR 11836 at commit [`5ea8469`](https://github.com/apache/spark/commit/5ea8469aafd347a7d1e69077de8d31a8f0167b25).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53900/
Test FAILed.

SparkQA commented:
**[Test build #53906 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53906/consoleFull)** for PR 11836 at commit [`9519cd8`](https://github.com/apache/spark/commit/9519cd806b1810e75283d4ca86fcb9aaf4d2d320).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53906/
Test FAILed.

yhuai commented:
Ah, we also need to update the python test after we changed the behavior of list tables in a non-existent db.

yhuai committed:
Use default as the db.

yhuai commented:
Looks like the conflict is introduced by https://github.com/apache/spark/pull/11889/files. Should be pretty minor. But, let me see.

yhuai commented:
looks like the conflicts are very easy to resolve.

yhuai commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/test/TestHive.scala:
```diff
@@ -83,26 +154,6 @@ class TestHiveContext(sc: SparkContext) extends HiveContext(sc) {
-    dir
-  }
-
-  private lazy val temporaryConfig = newTemporaryConfiguration(useInMemoryDerby = false)
```
I see why we have to remove 
```
TestHiveContext.hiveClientConfigurations(hiveconf, warehousePath, scratchDirPath)
  .foreach { case (k, v) => metadataHive.runSqlHive(s"SET $k=$v") }
```
in https://github.com/apache/spark/pull/11836/commits/16a54bad76a8297f15417ba931d20c4d86092c84. temporaryConfig was originally a lazy val. But, if we use hiveClientConfigurations, we are actually generate new confs every time we call reset.

yhuai commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/test/TestHive.scala:
```diff
@@ -83,26 +154,6 @@ class TestHiveContext(sc: SparkContext) extends HiveContext(sc) {
-
-  /** Sets up the system initially or after a RESET command */
-  protected override def configure(): Map[String, String] = {
-    super.configure() ++ temporaryConfig ++ Map(
```
Looks like as long as we make `temporaryConfig` a constant value, it is fine to have 
```
TestHiveContext.hiveClientConfigurations(hiveconf, warehousePath, scratchDirPath)
  .foreach { case (k, v) => metadataHive.runSqlHive(s"SET $k=$v") }
```
in the reset. I feel reset those confs is still needed just in case we override any of those confs.

Commit:
Merge branch 'master' of github.com:apache/spark into use-session-catalog

SparkQA commented:
**[Test build #53952 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53952/consoleFull)** for PR 11836 at commit [`c53f483`](https://github.com/apache/spark/commit/c53f483266475943e0fc5ac82ca28d973c4ff6f5).

Commit:
Fix R test?

SparkQA commented:
**[Test build #53955 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53955/consoleFull)** for PR 11836 at commit [`ea1645b`](https://github.com/apache/spark/commit/ea1645bd7cc60ffdf0b7fd2e9ea5db679ec2b9da).

SparkQA commented:
**[Test build #53956 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53956/consoleFull)** for PR 11836 at commit [`350bd2e`](https://github.com/apache/spark/commit/350bd2ebc2b44a5e6d3f58325aa9272f13201af6).

SparkQA commented:
**[Test build #53952 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53952/consoleFull)** for PR 11836 at commit [`c53f483`](https://github.com/apache/spark/commit/c53f483266475943e0fc5ac82ca28d973c4ff6f5).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53952/
Test FAILed.

SparkQA commented:
**[Test build #53955 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53955/consoleFull)** for PR 11836 at commit [`ea1645b`](https://github.com/apache/spark/commit/ea1645bd7cc60ffdf0b7fd2e9ea5db679ec2b9da).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53955/
Test PASSed.

rxin commented:
Merging in master - thanks!

SparkQA commented:
**[Test build #53956 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53956/consoleFull)** for PR 11836 at commit [`350bd2e`](https://github.com/apache/spark/commit/350bd2ebc2b44a5e6d3f58325aa9272f13201af6).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53956/
Test FAILed.
